substrate/emissions/pallet/src/lib.rs

```diff
@@ -0,0 +1,413 @@
+        // calculate distance to economic security per network
+        for n in NETWORKS {
+          if n == NetworkId::Serai {
+            continue;
+          }
+
+          let required = ValidatorSets::<T>::required_stake_for_network(n);
+          let mut current = ValidatorSets::<T>::total_allocated_stake(n).unwrap_or(Amount(0)).0;
+          if current > required {
+            current = required;
+          }
+
+          let distance = required - current;
+          distances.insert(n, distance);
+          total_distance = total_distance.saturating_add(distance);
+        }
+
+        // add serai network portion(20%)
+        let new_total_distance =
+          total_distance.saturating_mul(10) / (10 - SERAI_VALIDATORS_DESIRED_PERCENTAGE);
+        distances.insert(NetworkId::Serai, new_total_distance - total_distance);
+        total_distance = new_total_distance;
+
+        if Self::initial_period(n) {
+          // rewards are fixed for initial period
+          block_count * INITIAL_REWARD_PER_BLOCK
+        } else {
+          // rewards for pre-economic security is
+          // (STAKE_REQUIRED - CURRENT_STAKE) / blocks_until(SECURE_BY).
+          let block_reward = total_distance / Self::blocks_until(SECURE_BY);
+          block_count * block_reward
+        }
+      } else {
+        // post ec security
+        block_count * REWARD_PER_BLOCK
+      };
+
+      // get swap volumes
+      let mut volume_per_network: BTreeMap<NetworkId, u64> = BTreeMap::new();
+      for c in COINS {
+        // this should return 0 for SRI and so it shouldn't affect the total volume.
+        let current_volume = Dex::<T>::swap_volume(c).unwrap_or(0);
+        volume_per_network.insert(
+          c.network(),
+          (*volume_per_network.get(&c.network()).unwrap_or(&0)).saturating_add(current_volume),
+        );
+      }
+
+      // map current volumes to epoch volumes
+      let mut total_volume = 0u64;
+      for (n, vol) in &mut volume_per_network {
+        let last_volume = Self::last_swap_volume(n).unwrap_or(0);
+        let vol_this_epoch = vol.saturating_sub(last_volume);
+
+        // update the current volume
+        LastSwapVolume::<T>::set(n, Some(*vol));
+
+        total_volume = total_volume.saturating_add(vol_this_epoch);
+        *vol = vol_this_epoch;
+      }
+
+      // map epoch ec-security-distance/volume to rewards
+      let rewards_per_network = if pre_ec_security {
+        distances
+          .into_iter()
+          .map(|(n, distance)| {
+            // calculate how much each network gets based on distance to ec-security
+            let reward = u64::try_from(
+              u128::from(reward_this_epoch).saturating_mul(u128::from(distance)) /
+                u128::from(total_distance),
+            )
+            .unwrap();
+            (n, reward)
+          })
+          .collect::<BTreeMap<NetworkId, u64>>()
+      } else {
+        volume_per_network
```

Comment:
If we only use volume_per_network post-EC (as sane), we should only calculate it in this branch.

One `if` with two large branches, not two `if`s.